Add network functionality description for state and transaction gossip (#58)
ethereum/stateless-ethereum-specs: portal-network.md

### State: Accounts and Contract Storage

> The goal of the state network is to make pieces of the state available to light clients before the next block is mined.

Author: makes sense, updated!

> The state network needs to make pieces of the state available to light clients before the next block is mined, so that they can be quickly retrieved by the users' wallets and dapps.

Reviewer: ```suggestion
The state network facilitates on-demand retrieval of the Ethereum "state" data.
```

Nodes should be able to choose how much state they want to store and share, and the network should provide a way to identify which nodes to query for a wanted portion of state. This is so that every node, no matter how small, can contribute to the health and robustness of the network.

> The network needs to have a way to push state from new blocks into the network, so that it can be made available for query and retrieval by interested nodes. Full "bridge" nodes acting as benevolent state providers would be responsible for bringing in this data from the main network, and the network should be able to remain healthy even with a small number of bridge nodes.

Reviewer: ```suggestion
The network will be dependent on receiving new and updated state from new blocks. Full "bridge" nodes acting as benevolent state providers would be responsible for bringing in this data from the main network. The network should be able to remain healthy even with a small number of bridge nodes.
```

Slight re-wording since "the network" doesn't actually pull this data in so I wanted to change phrasing a little to make it clear that this isn't something the network is doing itself.  Also broke up the sentences a little.

Author: Makes sense! I also changed `from new blocks` to `for new blocks` as it might be interpreted that blocks are somehow pushing the state. I guess `of new blocks` might also work, considering the state root is in the header? Up to you!

### Transaction Sending: Cooperative Transaction Gossip

The goal of the transaction gossip network is to make sure all new transactions are made available to the miners so that they can be included in a block.

> Stateless clients should be able to declare how many transactions they want to process out of the set of all unmined and valid transactions (called _mempool_) based on the amount of resources they have, and should only receive that many transactions from other nodes.

Reviewer: I would change the phrasing minorly here.  A literal reading of this might leave the interpretation that participants are configuring *exactly* how many transactions they will be receiving.  In reality, they are tuning what *percentage* of the total mempool they will receive.

> Stateless transaction validation involves checking accounts' balances and nonces, so transactions should be broadcast with those proofs included.

Reviewer: ```suggestion
Stateless transaction validation involves checking accounts' balances and nonces, so the network will need to facilitate transmission of account proofs alongside each transaction.
```

### Transaction Sending: Cooperative Transaction Gossip

The goal of the transaction gossip network is to make sure all new transactions are validated and broadcast to the miners so that they can be included in a block.

> Stateless clients should be able to declare where they stand on the spectrum of "how many transactions do I want to process out of the whole mempool", based on the amount of resources they have and their role in the network, and the network should let clients filter and select destination nodes based on which transactions they are interested in.

Reviewer: > the network should let clients filter and select destination nodes based on which transactions they are interested in.

I'm not sure what this is trying to say.

Stateless transaction validation involves checking accounts' balances and nonces, so transactions should be broadcast with those proofs included.

> Transaction broadcasting involves pushing new transactions to the "center" of the network, where the miners are.  Note how this is different from the needs of the state network, where pieces of the state need to be "pushed out" from the full nodes to the light ones who will store it so they can participate in achieving or strenghening consensus.  This means that the "lightest" clients would not have to know about all (or any) transactions, they just need to make sure their own transactions are validated and sent to the miners as quickly as possible.

Reviewer: I think references to the "center" of the network get too close to implementation details.  While we're pretty sure that the network will be structured this way, it's not actually part of the requirements.  The more generic requirements are that:

1. Nodes can "tune" how much of the mempool they are required to process
2. Nodes can perform nonce and balance validation checks on transactions without a local copy of the state
3. Nodes can "tune" their network participation such that they receive the full mempool of transactions.

I think (2) is not yet captured in your description.
